README.md
# Painter

Live theme editing and UI customisation, brought to you by `Painter`.

- Change syntax / UI element CSS rules
- Tweak a single colour everywhere with fill mode
- Fine tune colour choice with modifiable colour sliders.

## Examples

#### Syntax

![](https://i.imgur.com/NX0lGFE.gif)

#### Syntax (Fill Mode)

![](https://i.imgur.com/zWyoU2N.gif)

#### UI

![](https://i.imgur.com/MwvYm9f.gif)

#### UI (Fill Mode)

![](https://i.imgur.com/rDz7x0L.gif)

## How to use

Commands:

`option-shift-p :: Paint Element`

`option shift-o :: Paint element (Fill mode)`

***Press command and then left click the element to paint***

Alternatively right click element to paint and select command from context menu.

- Adjust colour sliders
- Click outside window / press `esc` to cancel
- Press `Enter` to save

Painter applies live styles to your window, and saves final choices to your stylesheet (Does not affect existing styles).

To undo an accidental save, simply remove the style from your styles.less file, and reload the window.

## Installation

`apm install painter`

or directly through Atom.

\
\
Let me know any issues/suggestions/feature requests!

\- Keldan


CHANGELOG.md
## 0.0.0

Initial package launch


lib/utils.js
'use babel';


// Convert event to svg space
export function toSVG(event, refSVG) {
  while (refSVG.tagName != 'svg') refSVG = refSVG.parentElement;
  let p = refSVG.createSVGPoint();
  p.x = event.x;
  p.y = event.y;
  let newP = p.matrixTransform(refSVG.getScreenCTM().inverse());
  if (event.width && event.height) {
    let dims = toSVG({
      x: event.x + event.width,
      y: event.y + event.height}, refSVG);
    newP.width = dims.x - newP.x;
    newP.height = dims.y - newP.y;
  }
  return newP;
}

// Add attributes to svg elements
export function setAtts(element, values) {
  for (const val of Object.keys(values)) {
    element.setAttribute(val, values[val]);
  }
}

// Make svg element
export function make(tag, ...classes) {
  let elem = document.createElementNS("http://www.w3.org/2000/svg",tag);
  classes.forEach(c => elem.classList.add(c));
  return elem;
}

// Seperate check for element identifiers.
function matchesElem(identifier, element) {
  let checks = identifier.split(/([.#[])/).filter(Boolean);
  while (checks.length) {

    // Class
    if (checks[0] === ".") {
      if (!element || !((element.hasAttribute("class")
        && element.getAttribute("class").split(" ").includes(checks[1])))) {
        return false;
      }
      checks.shift();

    // ID
    } else if (checks[0] === "#") {
      if (!(element.getAttribute("id") === checks[1])) return false;
      checks.shift();

    // Attribute
    } else if (checks[0] === "[") {
      let attList = checks[1].slice(0,-1).split(/([~|^$*]?)=/).filter(Boolean);
      let val = element.getAttribute(attList[0]);
      let atlen = attList.length;

      // Case-insensitive
      if (attList[atlen-1].slice(-2) === " i") {
        attList[atlen-1] = attList[atlen-1].toLowerCase().slice(0,-2);
        val = val.toLowerCase();
      }
      // Lots of attribute options to check
      if ((atlen == 1 && !val)
        || (atlen == 2 && val !== attList[1])
        || (atlen == 3 && (
        (attList[1] === "~" && !val.split(/\s+/).includes(attList[2]))
          || (attList[1] === "|" && !(attList[2] === val
            || val.indexOf(attList[2] + "-") == 0))
          || (attList[1] === "^" && val.indexOf(attList[2]) != 0)
          || (attList[1] === "$" && val.slice(-attList[2].length) !== attList[2])
          || (attList[1] === "*" && !val.includes(attList[2]))
        ))) {
        return false;
      }

      // Move to next check
      checks.shift();
    } else if (element.tagName.toLowerCase() !== checks[0].toLowerCase()) {
      return false;
    }
    checks.shift();
  }
  // All checks complete
  return true;
}

// Check if css identifier applies to element
export const css = {
  "match": (rule, elem) => {
    var res = [];
    if (rule[0] === '@') return false; // Probably don't need to deal with these
    var sel, combs, elemPointers;

    // Get a list of conditions
    let selectors = rule.slice(0, rule.indexOf("{")).trim().split(/\s*,\s*/);
    for (let i = 0; i < selectors.length; i++) {
      elemPointers = [elem];
      sel = selectors[i];
      combs = sel.split(/\s*([\s>~+])\s*/).reverse().filter(Boolean);
      while (combs.length) {
        if (combs[0] === "*") {
          combs.shift();
          continue;
        }

        // Ancestor conditions
        if ([' ', '~'].includes(combs[0])) {
          let relation = (combs[0] === ' ')
            ? "parentElement" : "previousElementSibling";
          elemPointers = elemPointers.map(v => {
            let newElems = [];
            let par = v;
            while ((par = par[relation])) {
              if (matchesElem(combs[1], par)) newElems.push(par);
            }
            return newElems;
          }).flat();

        // Sibling conditions
        } else if (['>', '+'].includes(combs[0])) {
          let relation = (combs[0] === ' ')
            ? "parentElement" : "previousElementSibling";
          elemPointers = elemPointers.map(v => v[relation]).filter(v => {
            return (v && matchesElem(combs[1], v));
          });
        } else {
          elemPointers = elemPointers.filter(v => matchesElem(combs[0], v));
          combs.shift();
          continue;
        }
        combs.splice(0, 2);
      }
      if (elemPointers.length) {
        res.push(sel);
      }
    }
    return res;
  },

  // This probably doesn't need to be here
  "format": (rule, oldCol, newCol) => rule
    .replace(new RegExp(oldCol.replace(/[()]/g, '\\$&'), 'g'), newCol)
}


lib/slider.js
'use babel';

import {toSVG, setAtts, make} from './utils'

// Static id number for gradients
var id = 0;

export default class Slider {

  // Build slider
  constructor(parent) {
    this.sliderX = 0;
    let gradID = "painterSlider" + (id += 1);
    this.color = [0, 0, 0];

    // Make the elements
    this.elem = make("g", "slider");
    let bar = make("rect", "slidebar");
    let grad = make("linearGradient");
    let endStop = make("stop");
    this.handle = make("rect", "handle");

    // Set up some attributes
    setAtts(bar, {fill: `url(#${gradID})`});
    setAtts(grad, {id: gradID, x1: "0%", x2: "100%", y1: "50%", y2: "50%"});
    setAtts(endStop, {offset: "100%"});

    // Set up event listeners
    var handleClick = ((that) => (e => that.click(e)))(this);
    this.handle.addEventListener("mousedown", handleClick);
    bar.addEventListener("mousedown", handleClick);

    // Connect everything
    grad.appendChild(make("stop", "a"));
    grad.appendChild(endStop);
    this.elem.appendChild(bar);
    this.elem.appendChild(this.handle);
    this.elem.appendChild(grad);
    parent.element.appendChild(this.elem);

    // Set up some values we will need later
    this.parent = parent;
    this.handleOffset = Number(window.getComputedStyle(this.handle)
      .transform.match(/matrix\((?:.*?,){4}\s*?(.*?)\s*,/)[1]);
    this.handleWidth = Number(window.getComputedStyle(this.handle)
      .width.match(/([0-9.]*)/)[1]);
    this.barWidth = Number(window.getComputedStyle(bar)
      .width.match(/([0-9.]*)/)[1]);
    this.rgb = window.getComputedStyle(this.elem.children[2].children[1])
      .stopColor.slice(4, -1).split(", ").map(v => parseInt(v).toString(16));
  }

  // Destroy
  destroy() {
    if (this.clickX) this.release();
  }

  // Click and drag handling
  click(event) {
    this.dragHandler = ((that) => (e => that.drag(e)))(this);
    this.releaseHandler = ((that) => (() => that.release()))(this);
    document.addEventListener("mousemove", this.dragHandler);
    document.addEventListener("mouseup", this.releaseHandler);
    this.drag(event);
    this.clickX = toSVG(event, this.elem).x;
  }

  drag(event) {
    let newX = toSVG(event, this.elem).x - this.handleOffset - this.handleWidth/2;
    let maxX = Number(this.barWidth) - Number(this.handleWidth);
    let finalX = Math.max(Math.min(newX, maxX), 0);
    this.handle.setAttribute("x", finalX);
    this.changeColor(finalX/maxX);
    this.clickX = newX;
  }

  release() {
    document.removeEventListener("mousemove", this.dragHandler);
    document.removeEventListener("mouseup", this.releaseHandler);
    this.clickX  = null;
  }

  setValue(x) {
    let maxX = Number(this.barWidth) - Number(this.handleWidth);
    let newX = maxX*x/255;
    let finalX = Math.max(Math.min(newX, maxX), 0);
    this.handle.setAttribute("x", finalX);
    this.changeColor(finalX/maxX);
  }

  // Return element
  get element() {
    return this.elem;
  }

  // Compute new color based on handle position
  changeColor(factor) {
    this.color = this.rgb.map(v => (parseInt(v, 16)*factor>>0));
    let newColor = "#" + this.color.map(v=>v.toString(16))
      .map(v=>(v.length-1)?v:"0"+v).join("");
    this.handle.style.fill = newColor;
    this.parent.changeColor();
  }
}


lib/painter.js
'use babel';

import PainterView from './painter-view';
import { CompositeDisposable } from 'atom';
import fs from 'fs';
import path from 'path';

/* global atom */

export default {

  subscriptions: null,
  view: null,

  activate() {
    // Set up commands
    this.subscriptions = new CompositeDisposable();
    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'painter:paint': () => this.paint(),
      'painter:paint-fill': () => this.paint(true),
      'painter:paint-next': () => this.paintNext(),
      'painter:paint-next-fill': () => this.paintNext(true)
    }));

    // We need to capture the point at which the context menu is launched
    // That way we can get the element there
    this.listener = window.addEventListener("contextmenu",
      (event) => this.handleContextMenu(event));

    // Set up settings observer. We send slider colors to the stylesheet
    // We do that rather than just manage in in JS because I'm a lazy bitch.
    atom.config.observe('painter._sliderColors', nv => {
      let filePath = path.resolve(__dirname, './../styles/painter.less');
      let validColors = /(rgba?\((?:\d+,?\s*){3,4}\)|#[0-9a-f]{3,6})/;
      let vcList = nv.split(validColors).filter(f => validColors.test(f));
      if (vcList.length == 0) vcList = ["#ff0000", "#00ff00", "#0000ff"];
      fs.readFile(filePath, 'utf8',
        (err, data) => {
          if (err) throw err;
          let saveData = data.replace(/@bar-colors:.*?;/,
            `@bar-colors: ${vcList.join(', ')};`);
          if (saveData !== data) {
            fs.writeFile(filePath, saveData, err => { if (err) throw err;});
          }
        });
    });
  },

  // Main commands
  paint(isFill) {
    if (!this.lastClick) return;
    if (!this.view) this.view = new PainterView();
    this.view.trigger(this.lastClick, isFill);
  },

  paintNext(isFill=false) {
    this.isFill = isFill;
    this.awaitClick = ((that) => (e => that.waitForClick(e)))(this);
    document.addEventListener("mousedown", this.awaitClick);
  },

  // Keys have been pressed, waiting for click
  waitForClick(event) {
    this.lastClick = event;
    document.removeEventListener("mousedown", this.awaitClick);
    this.paint(this.isFill);
  },

  // We just need to store the event
  handleContextMenu(event) {
    this.lastClick = event;
  },

  deactivate() {
    this.subscriptions.dispose();
    window.removeEventListener(this.listener);
    this.view.destroy();
  },

  // Settings
  config: {
    _sliderColors: {
      title: 'Slider Colors',
      description: 'Set the color of the sliders (requires restart)',
      type: 'string',
      default: '#ff0000 #00ff00 #0000ff'
    }
  }
};


lib/painter-view.js
'use babel';

import fs from 'fs';
import path from 'path';
import Slider from './slider';
import {make, setAtts, css} from './utils';

/* global atom */

export default class PainterView {

  // Build
  constructor() {
    this._element = make("svg", "painter");

    // Add background
    this.background = make("rect", "background");
    setAtts(this.background, {fill: "#000000"});
    this._element.appendChild(this.background);

    // Attach to window
    atom.workspace.element.appendChild(this._element);

    // Add sliders
    let n_sliders = Number(window.getComputedStyle(this._element)
      .getPropertyValue("--sliders"));
    this.sliders = Array(n_sliders).fill().map(() =>new Slider(this));
  }


  // Summon the window
  trigger(event, isFillMode) {

    // Set fill mode
    this.isFillMode = isFillMode;

    // Prepare window for measurements
    setAtts(this._element, {display: "block", visibility: "hidden"});

    // Position window
    this.position(event.x, event.y);

    // Clicking outside element will close (and cancel) it
    this.closeCheck = ((that) => (e => that.checkIfClosing(e)))(this);
    document.addEventListener("mousedown", this.closeCheck);

    // Escape will cancel, enter/return will save
    this.buttonCheck = ((that) => (e => {
      if (["Enter", "Escape"].includes(e.key)) {
        that.endPaint(e.key === "Escape");
        e.stopPropagation();
      }
    }))(this);
    this.activeView = atom.views.getView(atom.workspace);
    this.activeView.addEventListener("keydown", this.buttonCheck);

    // Make element visible
    setAtts(this._element, {visibility: "visible"});

    // Get the target Element
    this.newColor = null;
    this.color = this.matchColor(event.path);
    this.originalColor = this.color;
    this.setSliders(this.originalColor);
  }

  // Handle clicks outside window while open
  checkIfClosing(e) {
    let wb = this._element.getBoundingClientRect();
    let inBox = (
      e.x >= wb.x
      && e.x <= wb.x + wb.width
      && e.y >= wb.y
      && e.y <= wb.y + wb.height
    );
    if (!inBox) this.endPaint(true);
  }

  // Apply Changes
  endPaint(cancelled=false) {
    if (!cancelled && this.newColor) {

      // Get user styles file
      let filePath = path.resolve(__dirname, './../../../styles.less');
      fs.readFile(filePath, 'utf8',
        (err, data) => {
          if (err) throw err;
          let saveData = this.formatStyle(data);
          fs.writeFile(filePath, saveData, err => { if (err) throw err;});
        });

    // If cancelled, revert
    } else this.changeColor(this.originalColor);

    // Clean up
    document.removeEventListener("mousedown", this.closeCheck);
    this.activeView.removeEventListener("keydown", this.buttonCheck);
    this.hide();
  }

  // Set sliders to current color
  setSliders(color) {
    let sliderCols = this.sliders.map(v => v.rgb.map(w => parseInt(w, 16)));

    // TODO: Add pattern recognition for other color formats
    let elemCol = color.match(/rgb\((\d*), (\d*), (\d*)\)/);
    elemCol = [elemCol[1], elemCol[2], elemCol[3]].map(v => Number(v));
    for (let s = 0; s < sliderCols.length; s++) {
      let val = Math.min(...sliderCols[s].map((v, i) => v ? elemCol[i]*255/v : Infinity), 255);
      elemCol = elemCol.map((v, i) => v - val*sliderCols[s][i]/255);
      this.sliders[s].setValue(val);
    }
  }

  // Save to styles.less
  formatStyle(content) {
    let header = /\s*\/\* Painter Styles \*\/\s*(.*?)\s*\/\* End Painter Styles \*\//ms;
    let innerText = "";
    let rules = {};

    // Parse each matching rule
    this.targetSheets.forEach(v => {
      let cssText = v.sheet.cssRules[v.index].cssText;
      let attributes = cssText.matchAll(
        new RegExp(`[;{]\\s*(\\S*?:[^:]*?${
          this.color.replace(/[()]/g, '\\$&')}.*?)(?=;)`, 'gm')
      );
      for (const match of attributes) {
        let attribute = match[1];
        if (!rules[attribute]) rules[attribute] = [];
        let newEntry = cssText.match(/([^{]*[^\s{])\s*/)[1];
        if (!rules[attribute].includes(newEntry)) {
          rules[attribute].push(newEntry);
        }
      }
    });
    // Parse relevant rules from existing data in user styles
    if (header.test(content)) {
      // Split Painter styles into rules
      const preStyles = RegExp.$1.split("\n").filter(Boolean);
      content = content.replace(header, '');
      preStyles.forEach(v => {
        let accessors = v.match(/.*?(?=\s*{)/)[0].split(/\s*,\s*/).filter(Boolean);
        let inner = v.match(/(?<={\s*)[^\s].*?(?=;\s*})/)[0];
        let innerRep = inner.replace(this.originalColor, this.color);
        if (!rules[innerRep]) rules[inner] = accessors;
        else {
          accessors.forEach(a => {
            if (!rules[innerRep].includes(a)) {
              if (!rules[inner]) rules[inner] = [];
              rules[inner].push(a);
            }
          });
        }
      });
    }
    // Finally build the file and save
    Object.keys(rules).forEach(v => innerText +=
      `${rules[v].join(', ')}{ ${v}; }\n\n`);
    return `${content}\n\n/* Painter Styles */\n\n${innerText}/* End Painter Styles */`;
  }



  // Position window
  position(x, y) {
    let screenHeight = document.body.clientHeight;
    let screenWidth = document.body.clientWidth;
    let windowBox = this._element.getBoundingClientRect();
    let yOffset = 25;

    // Don't let it go over the sides
    let left = Math.max(
      Math.min(screenWidth - windowBox.width, x - windowBox.width/2), 0
    ) + "px";

    // If window is too low, move it above mouse
    let top = ((y + yOffset + windowBox.height > screenHeight)
      ? y - yOffset - windowBox.height
      : y + yOffset) + "px";

    // Set element Position
    this._element.style.left = left;
    this._element.style.top = top;
  }

  // Hide window
  hide() {
    setAtts(this._element, {display: "none"});
  }

  // Called on changing slider value
  changeColor(color) {
    let colorVals = this.sliders.map(v=>v.color);
    let rC = colorVals.reduce((a,c)=>[a[0]+c[0], a[1]+c[1],a[2]+c[2]],[0,0,0])
      .map(v => Math.min(Math.max(0, v), 255));

    // Change background to match
    this.newColor = color || `rgb(${rC[0]}, ${rC[1]}, ${rC[2]})`;
    this.background.style.fill = this.newColor;

    // Apply color to target class in stylesheet or something
    this.targetSheets.forEach(v => {
      let rule = v.sheet.cssRules[v.index].cssText;
      let newCSS = css.format(rule, this.color, this.newColor);
      v.sheet.deleteRule(v.index);
      v.sheet.insertRule(newCSS, v.index);
    });
    this.color = this.newColor;
  }

  // Helper functions
  destroy() {
    this._element.remove();
  }

  get element() {
    return this._element;
  }

  // Find matching sheets
  matchSheets(color, elem) {
    let selectors;
    this.targetSheets = [];

    // Set up compare function for later
    let isMatch = cssText => {
      return cssText[0] !== '@' && cssText.includes(color) && (this.isFillMode
      ? true : (selectors = css.match(cssText, elem)).length)};

    // For each style in document
    for (let ss = 0; ss < document.styleSheets.length; ss++) {
      for (let i = 0; i < document.styleSheets[ss].cssRules.length; i++) {
        let cssText = document.styleSheets[ss].cssRules[i].cssText;
        if (isMatch(cssText)) {

          // Need to split apart some rules
          if (!this.isFillMode) {
            let ruleSelList = cssText.match(/.*?(?=\s*{)/)[0].split(/\s*,\s*/);
            let unaffected = ruleSelList.filter(v => !selectors.includes(v));
            let affected = ruleSelList.filter(v => selectors.includes(v));
            if (unaffected.length) {
              let newRule = cssText.replace(/.*?(?=\s*{)/, unaffected.join(', '));
              let otherNewRule = cssText.replace(/.*?(?=\s*{)/, affected.join(', '));
              let sheet = document.styleSheets[ss];
              sheet.deleteRule(i);
              sheet.insertRule(otherNewRule, i);
              sheet.insertRule(newRule, i);
              continue;
            }
          }
          // Store relevant rules
          this.targetSheets.push({
            sheet: document.styleSheets[ss],
            index: i
          })
        }
      }
    }
  }

  matchColor(path) {
    window.floop = path;
    let isText = path[0].tagName === 'SPAN'
      || path[0].childNodes[0].nodeType === 3;
    let matcher = isText
      ? /(?<={.*)(?<=\scolor:\s*)(?:rgba?\((?:\d+,?\s*){3,4}\)|#[0-9a-f]{3,6})/
      : /(?<={.*)(?<!\scolor:\s*)(?:rgba?\((?:\d+,?\s*){3,4}\)|#[0-9a-f]{3,6})/;
    let textColour = isText ? window.getComputedStyle(path[0]).color : null;
    for (let el = 0; el < path.length; el++) {
      let element = path[el];
      let sheets = document.styleSheets;

      // Iter backwards through all stylesheets
      // Since apparently they apply the last first
      for (let i = sheets.length - 1; i > -1; i--) {
        for (let j = sheets[i].cssRules.length -1; j > -1; j--) {
          let cssText = sheets[i].cssRules[j].cssText;
          if (textColour && !cssText.includes(textColour)) continue;
          if (css.match(cssText, element).length && matcher.test(cssText)) {
            let result = RegExp.lastMatch;
            this.matchSheets(result, element);
            return result;
          }
        }
      }
    }
  }
}
